src/custom-element.js

```diff
@@ -123,92 +123,103 @@ export function stubElements(win) {
  * @param {!AmpElement} element
  */
 export function applyLayout_(element) {
-  let widthAttr = element.getAttribute('width');
-  let heightAttr = element.getAttribute('height');
-  const sizesAttr = element.getAttribute('sizes');
   const layoutAttr = element.getAttribute('layout');
+  const widthAttr = element.getAttribute('width');
+  const heightAttr = element.getAttribute('height');
+  const sizesAttr = element.getAttribute('sizes');
+
+  // Input layout attributes.
+  const inputLayout = layoutAttr ? parseLayout(layoutAttr) : null;
+  assert(inputLayout || !layoutAttr, 'Unknown layout: %s', layoutAttr);
+  const inputWidth = widthAttr && widthAttr != 'auto' ?
+      parseLength(widthAttr) : widthAttr;
+  assert(inputWidth || !widthAttr, 'Invalid width value: %s', widthAttr);
+  const inputHeight = heightAttr ? parseLength(heightAttr) : null;
+  assert(inputHeight || !heightAttr, 'Invalid height value: %s', heightAttr);
+
+  // Effective layout attributes. These are effectively constants.
+  let width;
+  let height;
+  let layout;
 
-  // Handle elements that do not specify a width/height and are defined to have
-  // natural browser dimensions.
-  if ((!layoutAttr || layoutAttr == Layout.FIXED ||
-          layoutAttr == Layout.FIXED_HEIGHT) &&
-      (!widthAttr || !heightAttr) && hasNaturalDimensions(element.tagName)) {
+  // Calculate effective width and height.
+  if ((!inputLayout || inputLayout == Layout.FIXED ||
+          inputLayout == Layout.FIXED_HEIGHT) &&
+      (!inputWidth || !inputHeight) && hasNaturalDimensions(element.tagName)) {
+    // Default width and height: handle elements that do not specify a
+    // width/height and are defined to have natural browser dimensions.
     const dimensions = getNaturalDimensions(element.tagName);
-    if (layoutAttr != Layout.FIXED_HEIGHT) {
-      widthAttr = widthAttr || dimensions.width;
-    }
-    heightAttr = heightAttr || dimensions.height;
+    width = inputWidth || inputLayout == Layout.FIXED_HEIGHT ? inputWidth :
```

Comment:
This also might be better with parens:
width = inputWidth || (inputLayout == Layout.FIXED_HEIGHT ? inputWidth : dimensions.width);